Add html form pieces to support find_attitude >= 3.5.0 Constraints

jeanconn opened the pull request:
## Description

Add html form pieces to support find_attitude >= 3.5.0 Constraints.

## Interface impacts
<!-- API changes, file format updates, coordination of changes with the community. -->

## Testing
<!-- If relevant describe any special setup for testing. -->
Installed on kadi-test for functional testing.

### Unit tests
<!-- At least one of these must be checked if unit tests exist. DELETE the unchecked/untested options. -->
- [x] OSX
```
(2025.1-web) flame:kadi-apps jean$ pytest
============================================== test session starts ==============================================
platform darwin -- Python 3.12.8, pytest-8.3.4, pluggy-1.5.0
rootdir: /Users/jean/git
configfile: pytest.ini
plugins: anyio-4.7.0, timeout-2.3.1, dash-2.18.2
collected 26 items                                                                                              

kadi_apps/tests/test_astromon.py ....                                                                     [ 15%]
kadi_apps/tests/test_auth.py .......                                                                      [ 42%]
kadi_apps/tests/test_ska_api.py ...............                                                           [100%]

=============================================== warnings summary ================================================
kadi-apps/kadi_apps/tests/test_ska_api.py::test_starcheck_att
kadi-apps/kadi_apps/tests/test_ska_api.py::test_starcheck_att
  /Users/jean/miniforge3/envs/2025.1-web/lib/python3.12/site-packages/bs4/builder/_lxml.py:124: DeprecationWarning: The 'strip_cdata' option of HTMLParser() has never done anything and will eventually be removed.
    parser = parser(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================== 26 passed, 2 warnings in 11.42s ========================================
(2025.1-web) flame:kadi-apps jean$ git rev-parse HEAD
51d99e3ac35201c55c40e11bfd4866610301f147
```

### Functional tests
<!-- Describe and document results of any functional tests, otherwise leave the text below -->
The new form looks like this:

<img width="1259" alt="Screenshot 2025-01-10 at 11 21 12 AM" src="https://github.com/user-attachments/assets/bc6825a3-10a8-4448-abd6-4624d4f9dff8" />



And there's some new debug output that I found useful:
```
Date for Pitch/Roll: 2024:352:16:46:13.501
Sun Pitch: 149.9392937259457
Off-Nominal Roll: 7.276478482670257
```
```
Supplied find_attitude Inputs

slot   YAG      ZAG    MAG_ACA
---- -------- -------- -------
   3   669.65   144.72    7.25
   4  2454.88  -262.43    7.38
   5   978.08 -1594.53    8.31
   6 -1083.78   810.93     8.5
   7   -619.5    808.6    8.62 
 tolerance=2.5 
 constraints={}
find_attitude version: 3.5.1.dev0+g0b22b6b.d20241217
```

#### TA functional testing

All the tests below were OK.

- [x] During a maneuver in real time, the app reports no stars tracked during a maneuver.
- [x] Test with two stars during a normal observation where multiple solutions are found.
- [x] Test with three stars during a normal observation where multiple solutions are found.
- [x] Test various combinations of providing / not providing constraints.
- [x] Test gettings solutions in Safe Mode at 2023:046:00:34:15.230 with slots 1, 5, 6, 7.
   - No constraints and 3.0 arcsec dist tolerance finds no solution.
   - Sun pitch=90, pitch error=1.5, max_off_nom_roll = 2.0 matches 3 stars and gives solution.
   - Sun pitch=90, pitch error=1.5, max_off_nom_roll = 2.0, dist=4.0 arcsec matches 4 stars and gives solution.

jeanconn committed:
Add some of the find_attitude constraints to the form

jeanconn committed:
Sort imports

jeanconn committed:
Update some quotes

jeanconn committed:
More import reorg

jeanconn committed:
Add att and att_err constraints

jeanconn committed:
Update off nominal roll text

jeanconn committed:
Reorganize input/debug output

jeanconn commented:
I note that for me I found it a bit confusing that the default off nominal roll tolerance is pretty small, so if any constraint is supplied, that comes into play.  Not sure if we want to just set that larger from the form version.

jeanconn commented:
In our initial discussion it sounded like we also wanted this to also auto-fill the attitude from telemetry.

jeanconn committed:
Handle constraints separately and fill all from telem

jeanconn commented:
Per feedback in person, I updated the code to fill in pitch and attitude from telemetry.  It made more sense to me at that point to add an option to run find attitude with constraints and to run find attitude without constraints, so I rejiggered the buttons.

[Kadi.pdf](https://github.com/user-attachments/files/18324362/Kadi.pdf)

jeanconn committed:
Rearrange directions at top a bit more

taldcroft commented:
Just some labeling nitpicks first:
- `Optional Constraints to Limit Star Pair Candidates`
- `Pitch Uncertainty (degrees)`
- `Radial Uncertainty in Estimated Attitude (degrees)`
  -  Uncertainty in roll doesn't matter
-  `Min Required Stars (normally leave blank, discuss with ACA)`
- `Max Diff Between Observed - AGASC Mag`
  - Technically this only applies on the faint end but people get the idea.

jeanconn committed:
Set form to be two column

jeanconn commented:
Can you clarify "Uncertainty in roll doesn't matter"?  Does that mean you want to remove it?  By default it is the most limiting constraint in run-of-the-mill attitude testing, even if off nominal roll should be less during a real find_attitude situation.

I just liked having it so the user could change it.

jeanconn committed:
Update constraint labels per Tom

jeanconn committed:
Add information about solution vs est attitude

jeanconn committed:
Seed off nominal roll constraint and preserve date

taldcroft commented:
Roll is spacecraft roll (ra, dec, roll), not Off Nominal Roll. That comment was about the attitude uncertainty. For PCAD engineers, an error in roll counts as "attitude error", but that would make no difference here in finding the attitude.

jeanconn commented:
OK, so did you want this bullet "Uncertainty in roll doesn't matter" on the page somewhere to say uncertainty in roll does not matter with regard to the attitude uncertainty?

jeanconn committed:
Constraints need date

taldcroft commented:
No. Both of the sub bullets were just for you.

jeanconn commented:
I note one issue with the 2-column top is that the GRETA data doesn't copy-paste into the input box in a pretty fashion.
I don't think we care.

javierggt commented:
I was just testing kadi-apps on HEAD and on OSX using 2025.1rc1. Seems to work.

My first question is about which option combinations are allowed, and what happens if there are conflicting options. 
- I am assuming that one cannot give sun pitch and estimated attitude at the same time, is that true?
- Does pitch uncertainty work with estimated attitude?
- same with max off-nominal roll.

jeanconn commented:
With regard to the constraints, they can all be combined and from my perspective it is fine.  If you overconstrain the solution, you just don't get a solution.  

> I am assuming that one cannot give sun pitch and estimated attitude at the same time, is that true?

In my testing it seems fine to pass them both.

> Does pitch uncertainty work with estimated attitude?

I've just been passing everything but what test case would answer your question about if this "works"?

taldcroft commented:
@javierggt - each constraint is independent and serves to narrow down the list of candidate star pairs. So if you put in a conflicting constraints then there will be no stars and no solution.

In practice a conflict like this won't happen -- if the estimated attitude is considered reliable by the PCAD engineers then it will correspond to the Sun pitch.

It is a somewhat interesting question in this case (no stars meet all constraints) if the code will run to completion with no solution or if it will get an exception along the way. I don't recall that this has been tested.

javierggt commented:
ok, I was just asking as a user first confronted by the form. I expected some kind of grouping of options (like pitch and off-nominal roll) but they are not grouped.

So, in terms of grouping, sun pitch is right next to pitch uncertainty, so I infer that this pitch is actually sun pitch.

taldcroft commented:
@jeanconn - sorry for the churn, but I'd like a re-do on the labels. I had gone back and forth between "Estimated Attitude Radial Uncertainty" and "Radial Uncertainty for Estimated Attitude". Today when I look at the form, and considering @javierggt's experience, the former would be be better. Likewise, for consistency use `Sun Pitch Uncertainty (degrees)` and `Sun Pitch (degrees)`.

jeanconn committed:
More constraint label updates based on user feedback

jeanconn commented:
Labels updated!

jeanconn committed:
Update the no matching solutions text for optional constraints

taldcroft commented:
@jeanconn and/or @javierggt - can you make this PR live on the test server? Sorry for being lazy, but we're going to want to do this anyway before we finally approve the PR. (Testing on a local server does not guarantee that it will work on the live CXC server).

My preference, if it works, is to deploy this PR to the kadi web server with the current Ska environment. Then later think about upgrading the environment.

jeanconn committed:
Set default constraints conditionally

jeanconn committed:
Filter centroids that have MAG_ACA 13.94 and YAG/ZAG 3276.80

jeanconn committed:
Set printing relative to estimated attitude to be conditional in template

jeanconn committed:
Remove GRETA A_ACA_ALL example

jeanconn commented:
OK, in the latest changes I've

- Set the default max off nominal roll constraint back to 2 degrees instead of to a value based on the estimated attitude's off nominal roll.
- removed the GRETA example and doc
- fixed a bug where any of the new float constraints would cause everything to fail if not supplied.
- filtered centroids that are just obviously at the no-track state based on MAG_ACA and Y and Z (but this doesn't do anything clever)

Commit [3668eb3](https://github.com/sot/kadi-apps/pull/26/commits/3668eb3dd598f0d604265f7e94a4aa43b8fe98cc) is now live on kadi-test .

The outputs need a bit of cleanup as I've got too many digits in a bunch of the printing.

jeanconn commented:
On this PR feedback was I should have the solved quaternion available as comma-separated for easy pasting into SAUSAGE.

jeanconn committed:
Cut fid lights (AOACFID == 'FID') from auto-table

jeanconn committed:
Reformat all those output decimals and move quaternion to csv

jeanconn committed:
Floatify pitch and roll if needed

jeanconn committed:
Use roll0 instead of roll for delta

jeanconn commented:
If I get a "looks OK" for display content and smoke testing, I'll go through and document some more detailed functional testing.

javierggt commented:
> I've updated the kadi-test server with [35bb103](https://github.com/sot/kadi-apps/commit/35bb1038b2724211c952f9225947eb8145ce1ab0) but it looks like it has stopped displaying the right version at the bottom.

Did you restart apache? If not, there might be python instances running with the older version. It can happen that you get the newer version without restarting if a python instance happens to start.

jeanconn commented:
I restarted apache and it ended up printing a new/old version inconsistent with what I'd just pip installed.

jeanconn commented:
Sure, but I installed it with pip like so and the output matches the PR
```
[kadi@web-kadi-test kadi-apps]$ pip install .
Processing /home/kadi/git/kadi-apps
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: kadi-apps
  Building wheel for kadi-apps (setup.py) ... done
  Created wheel for kadi-apps: filename=kadi_apps-1.3.1.dev26+g35bb103-py3-none-any.whl size=208300 sha256=730c4bf7842ee68b11f8aa7b778f6890ffc9a1ab8f4a1431c4b27614f7072125
  Stored in directory: /home/kadi/.cache/pip/wheels/e2/62/94/381a1b3cf26e79686785c758ab2339e2c4f5a1649a8ffdbf04
Successfully built kadi-apps
Installing collected packages: kadi-apps
  Attempting uninstall: kadi-apps
    Found existing installation: kadi-apps 1.3.1.dev25+gc4895c7
    Uninstalling kadi-apps-1.3.1.dev25+gc4895c7:
      Successfully uninstalled kadi-apps-1.3.1.dev25+gc4895c7
Successfully installed kadi-apps-1.3.1.dev26+g35bb103
```

jeanconn commented:
```
[kadi@web-kadi-test ~]$ which pip
/export/servers/kadi-test/ska3/bin/pip
```

jeanconn commented:
I assume it is picking up the version from the directory
```
/export/servers/kadi-test/ska3/lib/python3.10/site-packages/kadi_apps-1.0.4.dev2+g71b282d.dist-info
```
even if that isn't what is running after the last pip install?  Though the fact it bounced around and displayed the dev25 version before going back to the ancient 1.0.4 is just weird.

jeanconn commented:
Support more gracefully when there is no star data like at time `2025:024:14:15:41.337` which has star data 
```
  slot        yag        zag     mag  
     0   -3276.80   -3276.80   13.94  
     1   -3276.80   -3276.80   13.94  
     2   -3276.80   -3276.80   13.94  
     3   -3276.80   -3276.80   13.94  
     4   -3276.80   -3276.80   13.94  
     5   -3276.80   -3276.80   13.94  
     6   -3276.80   -3276.80   13.94  
     7   -3276.80   -3276.80   13.94 
```
on the flight server.

jeanconn commented:
The PR should use AOACFCT to determine if a star is tracked and if none are tracked just print "No tracked stars" in the box.

jeanconn committed:
Change the order of the results so quats are last

jeanconn committed:
Use AOACFCT for track status

jeanconn committed:
Handle case of no tracked star data

jeanconn committed:
Add some error handling for bad quaternion

jeanconn committed:
Add more error handling on any constraint

jeanconn commented on kadi_apps/blueprints/find_attitude/find_attitude.py:
```diff
@@ -78,41 +128,113 @@ def index():
+        att_est_text = request.form.get("att", "").strip() or None
+        if att_est_text:
+            try:
+                att_est = Quat(q=[float(val) for val in att_est_text.split(",")])
```
Eh.  I added more exception handling that should be sufficient for the application.

jeanconn commented:
This needs my documentation of functional tests, but is ready for more beta testing and review.

jeanconn commented:
I also rebased so the test fixes are in here now too.

jeanconn commented:
Make it shorter so you can see the solution a bit on one page.

jeanconn commented:
Add a "generated at" time for Attitude Solution in local time.

jeanconn committed:
Add solution generation time to web page

jeanconn committed:
Tweak time range (helps for mismatch sample on quaternions and image data)

jeanconn committed:
Pretty-print constraints in input section

jeanconn committed:
Reduce vertical whitespace a bit

jeanconn committed:
Update whitespace and bulleted docs

jeanconn committed:
Add more error handling around missing telem

jeanconn commented:
And now this is throwing server errors I don't understand and the date I just asked it to print is not printing (both issues on kadi-test but not happening locally).  So this is going to be a bit longer.

jeanconn committed:
Explicitly use local timezone

jeanconn committed:
Update default distance tolerance to 3.0

jeanconn commented:
Todo:
Add centroid filtering/smoothing from https://github.com/sot/find_attitude/blob/master/find_attitude/find_attitude.py .  Also consider reducing to one function instead of duplicating the maude fetching in kadi-apps and find_attitude(we may have ended up with two functions to allow this to be installed more rapidly on kadi-test, not sure about the history here)

jeanconn committed:
Pass along constraints as None if blank

jeanconn committed:
Use generation time for Errors and solutions

jeanconn committed:
Tweak time ranges and add a little filter on yag/zag/mag

jeanconn committed:
Add arcsec to deltas

jeanconn committed:
Update test for the matching msids

jeanconn commented:
OK I implemented the comments from our in-person today including a "diving" filter on the data and a bit more stable handling of the input date.  Question @taldcroft what did you have in mind for the degrees/arcsecs formatting for the deltas?  Right now it is:

Distance from solution to onboard estimated attitude

dyaw=-0.00008 degrees (-0.3 arcsec)
dpitch=-0.00002 degrees (-0.1 arcsec)
droll=0.00172 degrees (6.2 arcsec)

jeanconn commented:
(and latest changes are live on the test server)

jeanconn committed:
Tweak times a bit more for stability

taldcroft commented:
Not necessarily a lien, but I did manage to get a 500 internal server error with these inputs, trying for a 2-star solution. I got telemetry for the given time and then just took two stars. It took about 20 seconds trying to respond and then gave a 500.

With this notebook, find_attitude finds 3 solutions:
https://gist.github.com/taldcroft/568f9c098bb1b5cc4cd2e35419b1fc8c

The web app gives a 500 with (I think) the same inputs:
<img width="1090" alt="image" src="https://github.com/user-attachments/assets/016e8315-72d1-46a1-a1cb-584444a766f5" />

Is it possible that multiple solutions are not being handled correctly in the app?

taldcroft commented:
This also gives a 500 in the app, whereas the notebook finds two solutions. It looks like multiple solutions is the problem.
<img width="1090" alt="image" src="https://github.com/user-attachments/assets/06f04e9c-0158-48d4-a2ea-4f8dbd52ed03" />

jeanconn commented:
Great testing and good catch @taldcroft!  

It looks like the issue from the kadi-apps code is that I'm expecting each "summary" in the solutions to basically be a table that has YAG, ZAG, and MAG, and it looks like the first solution is special and has those and the others do not (and the docstring just says there shall be dicts). I can certainly update the kadi-apps code to stop expecting YAG/ZAG/MAG, but is it consistent with your expectations that the summary has different content for the later solutions?

jeanconn committed:
Set format on YAG/ZAG/MAG_ACA only if in the table

jeanconn committed:
Add a line and a solution index between solutions

jeanconn commented:
I updated the kadi-apps code to work with these multiple solutions and added a horizontal line and a solution idx.

taldcroft commented:
Yup, I noticed that difference and was dismayed. Even worse, the output table for the 2nd, 3rd etc solutions does not include the slot, so doing a join on the input stars (with MAG, YAG, ZAG) is not possible. Ugh.

So I'm thinking about just catching this situation and dealing with it in some minimal way, and I'll fix find_attitude (eventually).

taldcroft commented:
Perfect, I confirmed that this works.

taldcroft reviewed (approved):
Awesome job @jeanconn! I have finished testing and documented my results in the description.
